src/include/network/peloton_rpc_handler_task.h

```diff
@@ -13,22 +13,135 @@
+#include "codegen/proxy/string_functions_proxy.h"
+#include "codegen/query.h"
+#include "codegen/query_cache.h"
+#include "codegen/query_compiler.h"
+#include "codegen/type/decimal_type.h"
+#include "codegen/type/integer_type.h"
+#include "codegen/type/type.h"
+#include "codegen/value.h"
+#include "planner/populate_index_plan.h"
+#include "traffic_cop/traffic_cop.h"
+#include "storage/storage_manager.h"
+#include "planner/seq_scan_plan.h"
 
 namespace peloton {
 namespace network {
 class PelotonRpcServerImpl final : public PelotonService::Server {
  protected:
-  kj::Promise<void> createIndex(CreateIndexContext) override {
-    // TODO(tianyu) Write actual index code
-    LOG_DEBUG("Received rpc to create index");
+  kj::Promise<void> dropIndex(DropIndexContext request) override {
+    auto database_oid = request.getParams().getRequest().getDatabaseOid();
+    auto index_oid = request.getParams().getRequest().getIndexOid();
+    LOG_DEBUG("Database oid: %d", database_oid);
+    LOG_DEBUG("Index oid: %d", index_oid);
+
+    auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    // Drop index. Fail if it doesn't exist.
+    auto catalog = catalog::Catalog::GetInstance();
+    try {
+      catalog->DropIndex(database_oid, index_oid, txn);
+    } catch (CatalogException e) {
+      LOG_ERROR("Drop Index Failed");
+      txn_manager.AbortTransaction(txn);
+      return kj::NEVER_DONE;
+    }
+    txn_manager.CommitTransaction(txn);
     return kj::READY_NOW;
   }
-};
 
+  kj::Promise<void> createIndex(CreateIndexContext request) override {
+    LOG_DEBUG("Received RPC to create index");
+    auto database_oid = request.getParams().getRequest().getDatabaseOid();
+    auto table_oid = request.getParams().getRequest().getTableOid();
+    auto col_oids = request.getParams().getRequest().getKeyAttrOids();
+    auto is_unique = request.getParams().getRequest().getUniqueKeys();
+    auto index_name = request.getParams().getRequest().getIndexName();
+    std::vector<oid_t> col_oid_vector;
+    LOG_DEBUG("Database oid: %d", database_oid);
+    LOG_DEBUG("Table oid: %d", table_oid);
+    for (auto col : col_oids) {
+      LOG_DEBUG("Col oid: %d", col);
+      col_oid_vector.push_back(col);
+    }
+
+    auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    // Create index. Fail if it already exists.
+    auto catalog = catalog::Catalog::GetInstance();
+    try {
+      catalog->CreateIndex(database_oid, table_oid, col_oid_vector,
+                           DEFUALT_SCHEMA_NAME, index_name, IndexType::BWTREE,
+                           IndexConstraintType::DEFAULT, is_unique, txn);
+    } catch (CatalogException e) {
+      LOG_ERROR("Create Index Failed");
+      txn_manager.AbortTransaction(txn);
+      return kj::NEVER_DONE;
+    }
+
+    // Index created. Populate it.
```

Comment:
Wouldn't create_index() in Catalog also populate tuples into it?  @mengranwo 

-- Tianyu, Justin & Tianyi

Reply:
No, it doesn't.